Run dashboard out-of-process

drewnoakes opened the pull request:
Fixes #1003
Fixes #851

This change causes the dashboard to be run as an orchestrated project, just like any other Aspire resource.

To achieve this, we:

- Break the project dependency from AppHost upon Dashboard.
  - Until we have support for dotnet tool resources, having the sample projects depend upon the dashboard directly, and register them with the DAB.
  - Copy `TimestampParser.DisplayFormat` into `Aspire.Hosting.Dashboard.ConsoleLogsConfigurationExtensions` where it was used, and add comments to keep them in sync.
- Have the dashboard wire up its own web application.
  - Remove `DashboardWebApplicationHost`.
  - Remove dashboard-specific code from `DistributedApplicationBuilder`.
  - Update URLs to resources defined in the dashboard. They no longer need (or work with) the `_content/Aspire.Dashboard/` prefix.
  - Remove workarounds/configuration to support static web assets, as the new structure supports this automatically.
  - Link `KnownProperties` and `KnownResourceTypes` into the hosting project, so both projects can share the definition without needing an assembly reference.
- Change the MSBuild SDK for the dashboard from Razor Class Library to Web project.
  - Remove explicit framework reference.
  - Remove redundant using directives (the new SDK has different implicit usings).
  - Make the dashboard an executable.
  - Update the CSS bundle file name (from `Aspire.Dashboard.bundle.scp.css` to `Aspire.Dashboard.styles.css`)

This work is incomplete for a few reasons:

1. Aspire users won't be able to create a project reference to the dashboard. That only works in this solution. We need support for resources defined via `dotnet tool` packages before we can fix that.
2. We don't launch a browser to the dashboard any more. You have to find the URL in the logs and launch it.

We will need to address those issues before the changes here can be merged.

---

To test this locally, run any of the sample applications. To find the dashboard's URL, navigate to `%TEMP%` and find the relevant log files. For example, `%TEMP%\aspire.oweozx1d.bmh\dashboard-5buv5so_out_eedbad51-0294-4ff8-9add-001f27499206` (your IDs will be different). The first log message should read _Now listening on: https://localhost:49337_ or similar. Open that URL in a browser.

---

Open questions before this can be marked as ready for review:

- [ ] We need DAM to support `dotnet tool`-based resources. When will that be available, and what APIs do we want?
- [ ] How do we preserve the development experience in this repo once we integrate dotnet tools? We want to be able to change the dashboard and hit F5.
- [ ] How to open a browser at the dashboard during launch in this model?
- [ ] Do we have to ensure the dashboard is running before other orchestrated processes start? We don't want to lose any OTEL data. See comments below about retry logic.

We should also consider to:

- [ ] Improve the experience when the resource service is unavailable.

---

## Screenshots

![image](https://github.com/dotnet/aspire/assets/350947/9da453b4-b85f-4958-8743-4af09bd08b86)

![image](https://github.com/dotnet/aspire/assets/350947/e48b53ab-6168-4236-8383-dca954407114)

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1667)

drewnoakes committed:
Run dashboard out-of-process

This change causes the dashboard to be run as an orchestrated project, just like any other Aspire resource.

To achieve this, we:

- Break the project dependency from AppHost upon Dashboard.
  - Until we have support for dotnet tool resources, having the sample projects depend upon the dashboard directly, and register them with the DAB.
  - Copy `TimestampParser.DisplayFormat` into `Aspire.Hosting.Dashboard.ConsoleLogsConfigurationExtensions` where it was used, and add comments to keep them in sync.
- Have the dashboard wire up its own web application.
  - Remove `DashboardWebApplicationHost`.
  - Remove dashboard-specific code from `DistributedApplicationBuilder`.
  - Update URLs to resources defined in the dashboard. They no longer need (or work with) the `_content/Aspire.Dashboard/` prefix.
  - Remove workarounds/configuration to support static web assets, as the new structure supports this automatically.
  - Link `KnownProperties` and `KnownResourceTypes` into the hosting project, so both projects can share the definition without needing an assembly reference.
- Change the MSBuild SDK for the dashboard from Razor Class Library to Web project.
  - Remove explicit framework reference.
  - Remove redundant using directives (the new SDK has different implicit usings).
  - Make the dashboard an executable.
  - Update the CSS bundle file name (from `Aspire.Dashboard.bundle.scp.css` to `Aspire.Dashboard.styles.css`)

This work is incomplete for a few reasons:

1. Aspire users won't be able to create a project reference to the dashboard. That only works in this solution. We need support for resources defined via `dotnet tool` packages before we can fix that.
2. We don't launch a browser to the dashboard any more. You have to find the URL in the logs and launch it.

We will need to address those issues before the changes here can be merged.

---

To test this locally, run any of the sample applications. To find the dashboard's URL, navigate to `%TEMP%` and find the relevant log files. For example, `%TEMP%\aspire.oweozx1d.bmh\dashboard-5buv5so_out_eedbad51-0294-4ff8-9add-001f27499206` (your IDs will be different). The first log message should read _Now listening on: https://localhost:49337_ or similar. Open that URL in a browser.

JamesNK commented:
> This change causes the dashboard to be run as an orchestrated project, just like any other Aspire resource.

A problem that occurs to me with treating the dashboard as a resource is the order things are loaded. Today the dashboard starts up at the same time other services are initialized. By the time orchestrated resources are loaded and OTLP data is sent to the dashboard, the dashboard is ready to receive it.

If the dashboard is loaded alongside other resources, then other resources could send OTLP data to the dashboard that isn't ready yet. Data could be lost.

I took a look at the OTEL SDK and there is infrastructure for retries, but nothing uses it yet: https://github.com/open-telemetry/opentelemetry-dotnet/pull/4616.

Testing is needed to figure out whether this is an actual problem.

adamint commented:
The problem of startup order might be solved by introducing the concept of initialization resources, similar to Kubernetes’ initContainers.

JamesNK commented:
Waiting until the dashboard is ready could solve the data loss problem, but if other resources must wait for the dashboard resource to startup, then it would increase total startup time of Aspire apps during development. Less would be happening in parallel and we would end up with:

Aspire host starts -> orchestration starts -> dashboard loads and is ready -> others projects start loading.

Do we care about the time between pressing F5 and seeing the dashboard UI? cc @DamianEdwards @davidfowl

kvenkatrajan commented:
> We need DAM to support dotnet tool-based resources. When will that be available, and what APIs do we want?
@mitchdenny  would you be able to assist with this please?

davidfowl commented:
We made some big decisions around how this will work today:
1. We're going to make the dashboard part of the workload instead of making it a .NET tool
2. We're going to teach the aspire host how to locate the executable at runtime (much like it does with finding DCP in the workload).
3. Aspire.Hosting will spin up the dashboard resource before other resources.
4. There will be no observable public API change to enable the dashboard outside of what exists for preview 2.

JamesNK commented:
> Aspire.Hosting will spin up the dashboard resource before other resources.

From a user perspective it looks like the DCP starts everything at the same time today. What does being first mean from a timing perspective?

If it is the first resource in a loop that starts all resources, won't that only offer a small difference (milliseconds) in start-up time?

JamesNK commented:
I created an issue with OTEL folks to enable retries - https://github.com/open-telemetry/opentelemetry-dotnet/issues/5226. I think that in the long term it is the best way to ensure OTEL data doesn't get lost.

drewnoakes committed:
Make dashboard packable as a dotnet tool

Also adds a readme explaining dashboard configuration, and sets some metadata on the package for the icon, readme, etc.

drewnoakes commented:
> From a user perspective it looks like the DCP starts everything at the same time today. What does being first mean from a timing perspective?
> 
> If it is the first resource in a loop that starts all resources, won't that only offer a small difference (milliseconds) in start-up time?

The idea was that the dashboard would be started and actual _readiness_ would be awaited before other resources were started. Readiness would be determined by probing OTEL somehow.

> I created an issue with OTEL folks to enable retries - [open-telemetry/opentelemetry-dotnet#5226](https://github.com/open-telemetry/opentelemetry-dotnet/issues/5226). I think that in the long term it is the best way to ensure OTEL data doesn't get lost.

Thanks. Agree that's the best way forward.

> We're going to make the dashboard part of the workload instead of making it a .NET tool

As far as I know, ACA will still want a dotnet tool for their container image. I've pushed an update to this PR that enables that.

drewnoakes committed:
Merge branch 'main' into dashboard-oop

mitchdenny commented on src/Aspire.Dashboard/DashboardWebApplication.cs:
```diff
@@ -7,45 +7,31 @@
-
         var builder = WebApplication.CreateBuilder();
+
         builder.Logging.AddFilter("Microsoft.Hosting.Lifetime", LogLevel.None);
```
We may not need these filters anymore since the dashboard isn't running in proc we don't need to worry about the noise it generates.

mitchdenny commented on src/Aspire.Dashboard/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## As a dotnet tool
+
+The dashboard can be packaged and executed as a `dotnet tool`.
```
This reads more like documentation for ourselves, but I think this `README.md` will end up being used as the README for the package, so perhaps we need to review the content in that context?

mitchdenny commented on src/Aspire.Hosting/Dashboard/DashboardExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+/// </summary>
+public static class DashboardExtensions
+{
+    public static IResourceBuilder<ExecutableResource> AddDashboard(this IDistributedApplicationBuilder builder)
```
We probably don't need this now (TODO below is out of date :P)

danegsta commented:
@drewnoakes do you want me to make the build targets changes required to integrate the dashboard into the workload in this branch? They shouldn't conflict with any of the current changes you've made.

mitchdenny commented:
@danegsta @drewnoakes why don't we get this PR merged onto another branch in the upstream so we can more easily collaborate there? Then we can merge the finalized branch in one go?

danegsta commented:
@mitchdenny @drewnoakes yeah, getting a feature branch going in the upstream makes sense.

davidfowl commented:
I dont think we should merge anything that breaks the workload into main. Why can't we commit on top of this branch?

joperezr commented:
I just created https://github.com/dotnet/aspire/tree/feature/dashboard-oop off from main. We can now retarget this PR to go against that branch, and I'll setup PR validation against that branch too.

joperezr commented:
Hey @drewnoakes, I took the liberty of retargeting your PR against a feature branch instead, this way we can hopefully get it in without disrupting main, and we can parallelize the rest of the dashboard-related work. I hope that's okay.

drewnoakes commented:
/azp run
